### DISCUSSION TITLE
Improvements to the simulator class (storing the output, holding the past input, simulating multiple steps and defining noise functions)

### DISCUSSION THREAD

**Discussion by Filippo-Tamagnini** [do-mpc/do-mpc#137] [Ideas]
**Is your feature request related to a problem? Please describe.**
1) The simulator class does not currently save the output/measurement in its data structure. This could be useful if do-mpc were to be used as a general modelling/simulation toolbox. 
2) Since the simulator class keeps u0 in memory, making it an optional parameter in the make_step() method would make sense (if the parameter is not passed, the simulator simply holds the past input).
3) It would be a nice feature to simulate multiple steps all at once, i.e: to generate first guesses for the MHE
4) Noise is currently set as an optional parameter in the simulator. I think it would make more sense to have a noise generating function explicitly defined, just as parameter generating functions are defined.

**Describe the solution you'd like**
1) I would modify the simulator class to store the measurement in the data structure. I think this would require some careful reasoning, though, as the y_next that is passed when calling make_step at t{k}, is actually the output at time t{k+1}. A perhaps inelegant way to do this would be to compute the measurement twice (before and after the integration and to store the first). Or we can simply say that the output is stored with 1 sample delay. This choice has huge implications, especially for discrete and/or semi-proper continuous systems.
2) Set u0=None as default parameter in simulator.make_step(...) and treat it in a similar way as the two noise parameters
3) Add N_steps parameter in simulator.make_step(...) and iterate. 
4) Add get_v_template and set_v_fun (idem for w) methods. Actually, thinking very hard about this, i'd move the set_p_fun, set_tvp_fun, bounds, normalization factors, etc (whatever is related to the model) into the model class.

**Additional context**
I have already started dirtifying the code, addressing points 1 and 2. Point 3 requires choosing a suitable output format for the make_step function (a list?). Point 4 requires some more careful consideration as well
here: 
https://github.com/Filippo-Tamagnini/do-mpc/tree/improvements-to-the-simulator-class-%23132

**Comment by Filippo-Tamagnini** [comment]
In addition to what i wrote yesterday, i have just realized there's an inconsistency between how the measurement/output is defined in the documentation (here: https://www.do-mpc.com/en/latest/api/do_mpc.model.Model.set_meas.html#do_mpc.model.Model.set_meas ) and how the simulator class behaves when stepping forward.
The documentation defines the measurement as **y[k] = g(x[k],u[k], ...)**, while the make_step method actually computes **y[k] = g(x[k+1],u[k], ... )**.

**Comment by 4flixt** [comment]
Hi Filippo, thanks again for the feedback!

1) This is indeed a bug - everything is prepared for the simulator to also store the measurements in the data structure (it exists and is initiated). I will raise an issue.

2) I think we could have ``u0``as an optional parameter for autonomous systems. In all other cases I would prefer to pass a value of ``u0``. This makes it clear to the user that an input is applied and which input is applied. I don't really see any advantage of just using the previous input again. 

3) Simulating multiple steps is typically done by just having this very simple loop:
```python
for k in range(steps):
   simulator.make_step(u0)
```
I would think that this negligible additional work for the user and most clear in what is happening. 
If you still think that this would be beneficial, we could check the shape of ``u0`` in ``simulator.make_step``
(it must currently be ``(n_u, 1)``. If it is of shape ``(n_u, steps)``, we could call ``make_step`` recursively.
It still feel this could lead to unexpected behaviour though and would only be marginally benefit the user. What do you think?

4) I guess that would be more consistent!

5) I will investigate this further, I think you have a point there.

As with your previous issue, I will move this to the discussions tab and add new issues regarding individual points.

Best
4flixt

**Reply by Filippo-Tamagnini** [reply]
>     2. I think we could have `u0`as an optional parameter for autonomous systems. In all other cases I would prefer to pass a value of `u0`. This makes it clear to the user that an input is applied and which input is applied. I don't really see any advantage of just using the previous input again.
It's just a matter of convenience imo. And if holding the previous input does not make sense in general, it would make sense to define some default action to be applied to the system, such as "do nothing".
>     3. Simulating multiple steps is typically done by just having this very simple loop:
Right, i'm just used to other simulator packages where you can ask to simulate steps in batches. And simulating multiple steps would probably conflict with the concept conveyed by the make_step statement

**Comment by 4flixt** [comment]
2) I really prefer the simulator to throw an error if there is no input. It just ensures that users will not get unexpected behaviour. Also I don't see any improvement in convenience. Like what is the difference between:
```python
# variant 1
for k in range(steps):
    simulator.make_step(u0)

# variant 2
simulator.make_step(u0)
for k in range(steps-1):
    simulator.make_step()
```
I don't think the code is clearer or easier to write.

3) We could simulate in batches but it will hardly make the code easier and under the hood everything will remain the same (we wont accelerate the code). I would probably opt to not change this
